### ISSUE TITLE
Make Certain Heavyweight Checkers Opt-In rather than Opt-Out

### ISSUE THREAD

**Issue by halostatue** [w0rp/ale#1337]
In a reply to a [comment](https://github.com/w0rp/ale/issues/826#issuecomment-320109931) on #826, a request was made to make Brakeman an opt-in rather than opt-out, the answer given was a flat 'no'.

I respectfully request that decision be reconsidered, or at least have an easy way to *know* what is running, or have a way of opting into some sort of heavy/light profile (so that checkers can self-identify as “heavy” and therefore not be in use by default). I recently upgraded ale and discovered that I was having numerous instances of `beam.smp` running, and it took me quite a while to figure out that there was a new linter added for Elixir that runs `dialyxer`.

You have said in other issues raised that you don’t know much about Elixir, but dialyzer has a lengthy (20+ minute) setup process on the first run for a given version of the BEAM VM (and possible a new version of Elixir). Because ALE’s policy is to run *all* available linters by default, and it launches new checking fairly frequently by default, my CPU was being pegged by ~30 instances of `mix dialyzer`…all trying to run the initial setup phase.

Other than upgrading (through vim-plug) ALE, I did not actually ask to run dialyzer on every run (and have since disabled it in a project-local vimrc). As you don’t know Elixir, you wouldn’t have known that this particular linter is *extremely* expensive to run (and cannot be run on an individual file, but only the whole project).

I don’t know what the right answer is for this problem, but short of removing dialyzer from my projects (which I don’t want to do), ALE will misbehave on Elixir projects simply because you added a new linter that has a high cost. To me, the best way would be to probably add some metadata:

```viml
call ale#linter#Define('elixir', {
\   'name': 'dialyxir',
\   'executable': 'mix',
\   'command': 'mix dialyzer',
\   'callback': 'ale_linters#elixir#dialyxir#Handle',
\   'cost': 3000, " or:
\   'weight': 'heavy',
\})
```

And then allow me to have something like `g:ale_linter_weight_profile = 'light'` or something like that. I still have to opt to choose the light profile, or choose the light profile per language, or something, but it would let me *not* be surprised by upgrading ALE and finding myself opted into a heavyweight linter.

**Comment by anandtrex** [comment]
I had the same problem with dialyzer with multiple processes consuming all my CPU. I just ended up disabling dialyzer in my vimrc (see below) and now just run dialyzer from the commandline when I need to.

```
let g:ale_linters = {
\ 'elixir': ['credo', 'dogma']
\ }
```

**Comment by halostatue** [comment]
Yes, that’s what I’ve done (although I don’t use dogma), but the problem here is that I got behaviour I don’t want simply by upgrading ale, and while the choice (run everything available) is probably good, making these choices without understanding the consequences for languages the maintainer doesn’t use…is IMO problematic.

I’m not asking for the default to change so much as I’m asking for:
* An easier way to see what linters are run for a file at any given time; this could have been solved faster if I could say something like `:ALEwhat` for the buffer and get an answer as to all the linters that can run.
* A way for linters to declare their impact. Not all linters are fast or can be run on a single file. Adding metadata to heavy linters (like dialyxer) would allow profiles to be built pretty easily and safely. The proposed `:ALEwhat` can even show what isn’t run because of profile exclusion, maybe. Or it might show something that allows the user to configure ALE more effectively for their use case.

I genuinely believe that the choice of running what can be run is sane. I don’t believe it should be as hard as it was to figure out what was killing my computer when I ran vim.

**Comment by halostatue** [comment]
@w0rp I get that ale runs everything it can without configuration, but I think you’re erring by closing this as `wontfix`, especially with the option to `let g:ale_linters_explicit` as your best possible recommendation (which it isn’t). If you *reread* what I wrote, I am not asking for the default to change. I have asked for two things:

1. More than anything, I want to be able to discover *what* linters will be run so that I can actually intelligently configure them on a project-by-project or global basis. What really irked me with the issue I had was that I had *no clue* that Dialyxir was added as a linter and all of a sudden, `ale` was bringing my system to its knees because it was running a very heavy linter as if it were a light linter. How can I ask `ale` what linters are going to run for a particular buffer? It *obviously* knows this, so why can’t I have it tell me? This helps me disable what I don’t want running.

2. What I want *secondarily* is to add metadata to the linters so that they can *self-declare* their weight. You yourself have said you don’t know everything about the linters you add. Let those of us who *do* know something about the linters in question add advisory metadata that can be used by people to improve their experience with `ale`. Leave the default for `ale` to run everything. Make it possible to filter based on metadata rather than explicit configuration.

At this point, I have to consider whether I trust `ale` as a lint-runner anymore based on your responses to people who have indicated that `ale`'s ability to DoS their system is problematic. I’m seriously leaning toward not.
